File: gpds/tests/testthat/test_dynamicalModel.R
library(testthat)
library(gpds)

testDynamicalModel <- function(dynamicalModelList, modelName, x, theta){
  context(paste(modelName, "model, with derivatives"))
  test_that(paste(modelName, "- Dx is consistent with f"), {
    f <- dynamicalModelList$modelODE(theta, x)
    
    deltaSmall <- 1e-6
    numericalDx <- sapply(1:ncol(x), function(j){
      xnew <- x
      xnew[,j] <- xnew[,j] + deltaSmall
      (dynamicalModelList$modelODE(theta, xnew) - f)/deltaSmall
    }, simplify = "array")
    
    fdX <- dynamicalModelList$modelDx(theta, x)
    
    expect_equal(fdX, aperm(numericalDx, c(1,3,2)), tolerance = 1e-4)
  })
  
  test_that(paste(modelName, "- Dtheta is consistent with f"), {
    f <- dynamicalModelList$modelODE(theta, x)
    
    deltaSmall <- 1e-6
    numericalDtheta <- sapply(1:length(theta), function(i){
      thetaNew <- theta
      thetaNew[i] <- theta[i] + deltaSmall
      (dynamicalModelList$modelODE(thetaNew, x) - f)/deltaSmall
    }, simplify = "array")
    
    fDtheta <- dynamicalModelList$modelDtheta(theta, x)
    
    expect_equal(fDtheta, aperm(numericalDtheta, c(1,3,2)), tolerance = 1e-4)
  })
}

x <- cbind(1:4, 4:1)
theta <- c(0.2, 0.2, 3)
dynamicalModelList <- list(
  modelODE = gpds:::fnmodelODE,
  modelDtheta = gpds:::fnmodelDtheta,
  modelDx = gpds:::fnmodelDx
)
testDynamicalModel(dynamicalModelList, "FN system", x, theta)

x <- cbind(1:4, 4:1, sin(1:4))
theta <- c(0.022, 0.3, 0.031, 0.028, 0.5, 20, 0.3)
dynamicalModelList <- list(
  modelODE = gpds:::hes1modelODE,
  modelDtheta = gpds:::hes1modelDtheta,
  modelDx = gpds:::hes1modelDx
)
testDynamicalModel(dynamicalModelList, "Hes1 system", x, theta)

File: MATLABmagi/testDynamicalModel.m
function passed = testDynamicalModel(modelODE, modelDx, modelDtheta, modelName, x, theta, tvec)
% Given functions for the ODE and its gradients (with respect to the system components and parameters), 
% verify the correctness of the gradients using numerical differentiation.
%
%      modelODE: function that computes the ODEs
%      modelDx: function that computes the gradients of the ODEs with respect to the system components
%      modelDtheta: function that computes the gradients of the ODEs with respect to the parameters theta
%      modelName: string giving a name for the model
%      x: data matrix of system values, one column for each component, at which to test the gradients
%      theta: vector of parameter values for theta, at which to test the gradients
%      tvec: vector of time points corresponding to the rows of x
% 
% RETURN A struct with fields Dx and Dtheta, with value 1 if the corresponding gradient check passed and 0 if not.
%  
    deltaSmall = 1e-06;
    tolerance = 1e-04;
    passed.modelName = modelName;
    
    f = modelODE(theta, x, tvec);
    fdX = modelDx(theta, x, tvec);
    
    numericalDx = NaN( length(tvec), size(x,2), size(x,2));
    for j=1:size(x,2)
        xnew = x;
        xnew(:,j) = xnew(:,j) + deltaSmall;
        numericalDx(:,j,:) = (modelODE(theta, xnew, tvec) - f) / deltaSmall;
    end
    
    passed.Dx = max( abs(numericalDx - fdX), [], 'all') < tolerance;
    
    fDtheta = modelDtheta(theta, x, tvec);
    
    numericalDtheta = NaN( length(tvec), length(theta), size(x,2));
    for j=1:length(theta)
        thetanew = theta;
        thetanew(:,j) = thetanew(:,j) + deltaSmall;
        numericalDtheta(:,j,:) = (modelODE(thetanew, x, tvec) - f) / deltaSmall;
    end
    
    passed.Dtheta = max( abs(numericalDtheta - fDtheta), [], 'all') < tolerance;
    
    
    
    
    
    
        
    
    